### ISSUE TITLE
bin2chen - liquidate() Lack of restriction that liquidator cannot be a CreditAccount

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-05-ironbank-judging#458]
bin2chen

medium

# liquidate() Lack of restriction that liquidator cannot be a CreditAccount

## Summary
in `liquidate()` liquidator can get `IBToken` through `marketCollateral`.
But the current method does not restrict liquidate from being `CreditAccount`.

## Vulnerability Detail
The protocol restricts `CreditAccount` from having an `IBToken` in several places

`supply()`
```solidity
    function supply(address from, address to, address market, uint256 amount)
        external
        nonReentrant
        isAuthorized(from)
    {
        DataTypes.Market storage m = markets[market];
        require(m.config.isListed, "not listed");
        require(!m.config.isSupplyPaused(), "supply paused");
@>      require(!isCreditAccount(to), "cannot supply to credit account");
```
`transferIBToken`
```solidity
    function transferIBToken(address market, address from, address to, uint256 amount) external {
        DataTypes.Market storage m = markets[market];
        require(m.config.isListed, "not listed");
        require(msg.sender == m.config.ibTokenAddress, "!authorized");
        require(!m.config.isTransferPaused(), "transfer paused");
        require(from != to, "cannot self transfer");
@>      require(!isCreditAccount(to), "cannot transfer to credit account");

        _accrueInterest(market, m);
        _transferIBToken(market, m, from, to, amount);

        _checkAccountLiquidity(from);
    }
```
But `liquidate()` can get IBToken, but there is no restriction
```solidity
    function liquidate(
        address liquidator,
        address borrower,
        address marketBorrow,
        address marketCollateral,
        uint256 repayAmount
    ) external nonReentrant isAuthorized(liquidator) {
        DataTypes.Market storage mBorrow = markets[marketBorrow];
        DataTypes.Market storage mCollateral = markets[marketCollateral];
        require(mBorrow.config.isListed, "borrow market not listed");
        require(mCollateral.config.isListed, "collateral market not listed");
        require(isMarketSeizable(mCollateral), "collateral market cannot be seized");
        require(!isCreditAccount(borrower), "cannot liquidate credit account");
        require(liquidator != borrower, "cannot self liquidate");

@>      _transferIBToken(marketCollateral, mCollateral, borrower, liquidator, ibTokenAmount);
        IBTokenInterface(mCollateral.config.ibTokenAddress).seize(borrower, liquidator, ibTokenAmount); // Only emits Transfer event.

        emit Liquidate(liquidator, borrower, marketBorrow, marketCollateral, repayAmount, ibTokenAmount);
```

Suggest adding restrictions

## Impact
Lack of restriction that liquidator cannot be a CreditAccount

## Code Snippet
https://github.com/sherlock-audit/2023-05-ironbank/blob/main/ib-v2/src/protocol/pool/IronBank.sol#L314

https://github.com/sherlock-audit/2023-05-ironbank/blob/main/ib-v2/src/protocol/pool/IronBank.sol#L566

## Tool used

Manual Review

## Recommendation

```solidity
    function liquidate(
        address liquidator,
        address borrower,
        address marketBorrow,
        address marketCollateral,
        uint256 repayAmount
    ) external nonReentrant isAuthorized(liquidator) {

        require(!isCreditAccount(borrower), "cannot liquidate credit account");
        require(liquidator != borrower, "cannot self liquidate");
+       require(!isCreditAccount(liquidator), "cannot transfer to credit account");
```

**Comment by ibsunhub** [comment]
fix: https://github.com/ibdotxyz/ib-v2/pull/55

**Comment by IAm0x52** [comment]
Fix looks good. Same fix as #98
